Add removeOne & removeMany functions

sangaman opened the pull request:
This adds functionality to remove multiple items from the queue with a single call. It works similar to the `remove` method. I will update the README in this PR so don't merge yet but I thought I would just check that the logic looks correct.

sangaman committed:
Add removeOne & removeMany functions

Restores `remove` as a method that removes a given item based on its
value.

Adds `removeOne` and `removeMany` methods that remove items based on
whether passing the item into a provided callback
function returns true.

Adds tests and updates README.

moshababo commented:
Looks good, but since we don't implement any clever removal code, and everything still relies on the callback (so the code is pretty much the same), maybe we can just add an optional parameter to `remove` on whether to break the scan after the first match, or to continue all the way and to return a counter. Then we can expose `removeOne` and `removeMany` on the top of it, in order to be more explicit publicly.

sangaman commented:
OK I did a deep dive into this. I found out that there actually *does* need to be some "clever" code around removing multiple items, simply iterating through the array doesn't always work because the array changes each time an item is removed. The solution I came up with is to do a bitwise shift on our current index of the array each time we remove an element, effectively this moves us "up" one level in the heap.

I restored the `remove` method to work like it did originally except without the possibility to pass a new comparator - this is a way to remove by value. I then created `removeOne` and `removeMany` which both use the same underlying internal method & logic and remove items according to custom logic provided in the callback. I also added tests and updated the README. `removeMany` takes an optional parameter to limit the number of items removed, and `removeOne` is essentially the same as calling `removeMany(callback, 1)`.

Commit:
removeMany tests

moshababo commented:
Looks great..
I played with it a bit and found an issue when having more then one item + trying to remove all items. I added the test case which currently fail so you'll see what I mean.

I'm gonna merge this regardless, because it currently blocks our progress in `xud` (naming and stuff).
